rfcs/0003-declaration-engine.md

```diff
@@ -0,0 +1,189 @@
+- Feature Name: Declaration Engine
+- Start Date: 2022-07-13
+- RFC PR: [FuelLabs/sway-rfcs#0011](https://github.com/FuelLabs/sway-rfcs/pull/11)
+- Sway Issue: [FueLabs/sway#1821](https://github.com/FuelLabs/sway/issues/1821), [FueLabs/sway#1692](https://github.com/FuelLabs/sway/issues/1692)
+
+# Summary
+
+[summary]: #summary
+
+With it's current design, the Sway compiler faces challenges regarding how declarations interact with the type system and code generation. These include:
+1. function bodies must be inlined during type checking ([FueLabs/sway#1557](https://github.com/FuelLabs/sway/issues/1557))
+2. monomorphization of declarations is unnecessarily duplicated ([FueLabs/sway#862](https://github.com/FuelLabs/sway/issues/862))
+3. it is currently impossible to implement `where` clauses without extensive special casing/dummy definitions ([FueLabs/sway#970](https://github.com/FuelLabs/sway/issues/970))
+
+This RFC proposes a solution to this---the "declaration engine". The declaration engine is a redesign of how the compiler thinks about type checking of declarations, and seeks to solve the problems described above. At a high level, the declaration engine stores intermediate representations of potentially type-checked declarations and allows the compiler to refer to those intermediate representations during type checking.
+
+# Motivation
+
+[motivation]: #motivation
+
+With the changes introduced by this RFC, the compiler will be able to think about declarations more abstractly and will not be required to inline AST nodes during type checking. In addition to solving the issues above, I believe that this change will create a mechanism in the compiler which will allow us to implement additional optimizations in the future.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Because this is an internal compiler RFC, the affected group is the Sway compiler devs.
+
+Specifically in the context of Sway terminology, the phrase "type checking" is often used to refer to the Sway compiler step in which untyped AST nodes are transformed to typed AST nodes. But really, the compier step we refer to "type checking" is actually doing type inference. This distinction is important because it allows us to understand the role of the declaration engine.
+
+## Toy Example
```

Comment:
Since this isn't really an example of what you are proposing, and is rather a description of a related concept, maybe we could title this section differently.
```suggestion
## Background Information: Sway's Type Inference
```